CRM/ACL/BAO/ACL.php

```diff
@@ -323,73 +327,68 @@ public static function group(
- WHERE   c.acl_id       =  a.id
-   AND   a.is_active    =  1
-   AND   a.object_table = %1
-   AND   a.id        IN ( $aclKeys )
-GROUP BY a.operation,a.object_id
-ORDER BY a.object_id
-";
-        $params = [1 => [$tableName, 'String']];
-        $dao = CRM_Core_DAO::executeQuery($query, $params);
-        while ($dao->fetch()) {
-          if ($dao->object_id) {
-            if (self::matchType($type, $dao->operation)) {
-              $ids[] = $dao->object_id;
+    else {
+      if ($allGroups == NULL && $tableName === 'civicrm_saved_search') {
+        $allGroups = CRM_Contact_BAO_Contact::buildOptions('group_id', 'get');
+      }
+      $acls = CRM_ACL_BAO_Cache::build($contactID);
+      $ids = [];
+      if (!empty($acls)) {
+        $aclKeys = array_keys($acls);
+        $aclKeys = implode(',', $aclKeys);
+
+        $cacheKey = CRM_Utils_Cache::cleanKey("$type-$tableName-$aclKeys");
+        $cache = CRM_Utils_Cache::singleton();
+        $ids = $cache->get($cacheKey);
+        if (!is_array($ids)) {
+          $ids = [];
+          $query = "
+  SELECT   a.operation, a.object_id
+    FROM   civicrm_acl_cache c, civicrm_acl a
+   WHERE   c.acl_id       =  a.id
+     AND   a.is_active    =  1
+     AND   a.object_table = %1
+     AND   a.id        IN ( $aclKeys )
+  GROUP BY a.operation,a.object_id
+  ORDER BY a.object_id
+  ";
+          $params = [1 => [$tableName, 'String']];
+          $dao = CRM_Core_DAO::executeQuery($query, $params);
+          while ($dao->fetch()) {
+            if ($dao->object_id) {
+              if (self::matchType($type, $dao->operation)) {
+                $ids[] = $dao->object_id;
+              }
             }
-          }
-          else {
-            // this user has got the permission for all objects of this type
-            // check if the type matches
-            if (self::matchType($type, $dao->operation)) {
-              foreach ($allGroups as $id => $dontCare) {
-                $ids[] = $id;
+            else {
+              // this user has got the permission for all objects of this type
+              // check if the type matches
+              if (self::matchType($type, $dao->operation)) {
+                foreach ($allGroups as $id => $dontCare) {
+                  $ids[] = $id;
+                }
               }
+              break;
             }
-            break;
           }
+          $cache->set($cacheKey, $ids);
         }
-        $cache->set($cacheKey, $ids);
       }
     }
 
-    if (empty($ids) && !empty($includedGroups) &&
-      is_array($includedGroups)
-    ) {
-      // This is pretty alarming - we 'sometimes' include all included groups
-      // seems problematic per https://lab.civicrm.org/dev/core/-/issues/1879
-      $ids = $includedGroups;
-    }
-    if ($contactID) {
+    // Contacts create hidden groups from search results. They should be able to retrieve their own.
+    if ($contactID && $tableName === 'civicrm_saved_search') {
```

Comment:
Added `&& $tableName === 'civicrm_saved_search'` to this condition because this block applies only to smart groups (I think)